### PR TITLE
Enhance Puppet Provisioning

### PR DESCRIPTION
**Pull request by gronke**
- [x] test puppet provisioner
- [x] refactor puppet provisioner
- [x] make is_path_relative() and require_relative_path() methods of JailGenerator public
- [ ] allow cloning manifests from remote sources without connecting the jail to any network

### CODE REVIEW

**Comment by gronke** [comment]
@igalic It should be possible to clone the manifest with the Python git module, so that no internet is required for this step. Can you imagine to create a super simple puppet manifest, that does something we can test. For example touch a file in the jails root directory? The current example manifest requires internet, which is a bit unfortunate in automated tests.

File: libioc/Provisioning/puppet.py [hunk 1]
```diff
@@ -169,91 +171,122 @@ def provision(
...
+    yield puppetApplyEvent.begin()
+    _, _, returncode = self.jail.exec([
+        "puppet",
+        "apply",
+        "--debug",
+        "--logdest",
+        "syslog",
+        f"{__puppet_env_dir}/production/manifests/site.pp"
+    ])
+    if returncode == 0:
```
Comment (igalic) [inline libioc/Provisioning/puppet.py]: https://puppet.com/docs/puppet/6.4/man/apply.html

we might wanna use `--detailed-exitcodes` here, to better be able to distinguish whether to rerun this, or not

File: libioc/Provisioning/puppet.py [hunk 2]
```diff
@@ -169,91 +171,122 @@ def provision(
...
         started = False
 
+    # Install packages
     pkg = libioc.Pkg.Pkg(
         logger=self.jail.logger,
         zfs=self.jail.zfs,
         host=self.jail.host
     )
-
+    yield from pkg.fetch(
```
Comment (igalic) [inline libioc/Provisioning/puppet.py]: i didn't realise that install doesn't fetch, but given the architecture it makes sense

File: tests/assets/puppet-control-repo/site/profile/lib/facter/ioc.rb [hunk 3]
```diff
@@ -0,0 +1,15 @@
+#
+# Add custom facts as passed on by (lib)ioc thru the execution environment
+#
+# Our facts are placed into user.facts.x which generates an environment variable
+# named IOC_USER_FACTS_X, which then is reduced to :x and its value.
+#
```
Comment (igalic) [inline tests/assets/puppet-control-repo/site/profile/lib/facter/ioc.rb]: this is is not used by our tests

File: tests/assets/puppet-control-repo/site/profile/manifests/base.pp [hunk 4]
```diff
@@ -0,0 +1,5 @@
+class profile::base {
+
+  #the base profile should include component modules that will be on all nodes
```
Comment (igalic) [inline tests/assets/puppet-control-repo/site/profile/manifests/base.pp]: this file is not used by our tests

File: tests/assets/puppet-control-repo/site/profile/manifests/example.pp [hunk 5]
```diff
@@ -0,0 +1,3 @@
+class profile::example {
+
```
Comment (igalic) [inline tests/assets/puppet-control-repo/site/profile/manifests/example.pp]: this file is not used by our tests

File: tests/assets/puppet-control-repo/hiera.yaml [hunk 6]
```diff
@@ -0,0 +1,12 @@
+---
+version: 5
```
Comment (igalic) [inline tests/assets/puppet-control-repo/hiera.yaml]: this file is not used by our tests

it, and all files in `data` can be omitted

File: tests/assets/puppet-control-repo/site/role/manifests/database_server.pp [hunk 7]
```diff
@@ -0,0 +1,7 @@
+class role::database_server {
+
+  #This role would be made of all the profiles that need to be included to make a database server work
```
Comment (igalic) [inline tests/assets/puppet-control-repo/site/role/manifests/database_server.pp]: this file is not used by our tests

the entire `role` dir can be omitted

**Comment by igalic** [comment]
> It should be possible to clone the manifest with the Python git module, so that no internet is required for this step.

yesno: `r10k` also fetches the required modules specified in the `Puppetfile` in the repository that is cloned 

> Can you imagine to create a super simple puppet manifest, that does something we can test. For example touch a file in the jails root directory? The current example manifest requires internet, which is a bit unfortunate in automated tests.

```puppet
file { '/foo':
  ensure => present,
  content => 'puppet test',
}
```

**Comment by gronke** [comment]
Thanks for looking into it, @igalic. Would you mind to apply the changes to this branch? I'll then fit the test, so that we can go ahead and add more specific tests.

**Pull request merged**